src/lib.rs

```diff
@@ -108,37 +125,45 @@ impl<IT, N> Slots<IT, N>
     }
 
     fn free(&mut self, idx: usize) {
-        self.free_list[self.free_count] = idx;
+        self.items[idx] = match self.next_free {
+            Some(n) => Entry::EmptyNext(n),
+            None => Entry::EmptyLast
+        };
+        self.next_free = Some(idx);
         self.free_count += 1;
     }
 
     fn alloc(&mut self) -> Option<usize> {
         if self.count() == self.capacity() {
             None
         } else {
-            let i = self.free_list[self.free_count - 1];
+            let result = self.next_free;
+            self.next_free = match self.items[result.expect("Count mismatch")] {
+                Entry::EmptyNext(n) => Some(n),
+                Entry::EmptyLast => None,
+                _ => unreachable!("Non-empty item in entry behind free chain"),
+            };
```

Comment:
I believe this makes the outer count check redundant. If next_free is None, there are no free slots to allocate from, so None can be returned.

Reply:
It was indeed, fixed in the next version.

I was mentally already in the next stage (only having EmptyNext with possible garbage value as in the "further optimization potential" section), but if and when that is used, it'll be sufficient to pull up the self.free_count decrement and make it `self.free_count = self.free_count.checked_sub(1)?;`.